### PR TITLE
WIP: fix memory leak & run crash

### PR DESCRIPTION
**Pull request by fucd**
* fix memory leak： memory monitor shows memory upexpected increase while run silicon tracking algorithm
  * [Utilities/KalTest/src/kaltracklib/TKalDetCradle.cxx]: A has object pointers Cs from B, but after transfer them, B's pointer is set as C's parent pointer. While delete A, only Cs are deleted, Bs become lost. So add delete B. But because some Cs has same parent, map is used to merge same B to avoid double free. B aslo has Cs, A also has Cs. In order to avoid double free, B.SetOwner(kFALSE) is needed to release Cs, after B transfer Cs into A.
  * [Service/TrackSystemSvc/src/MarlinTrkUtils.cc]: new TrackState should always be deleted, because track->addToTrackStates(edm4hep::TrackState) will create a TrackState object.
  * [Detector/GeomSvc/src/GeomSvc.cpp]: dd4hep::Detector::destroyInstance to free dd4hep objects.
  * [Digitisers/SimpleDigi/src/PlanarDigiAlg.cpp]: gsl_rng_free(_rng) to free gsl_rng.
* fix run crash:
  * DriftChamberRegion use PAI model will cause crash (old problem, fixed for CRD_o1_v01, but others are missed), remove this region definition preliminary same as CRD_o1_v01.
  * [Reconstruction/Tracking/src/Clupatra/ClupatraAlg.cpp]: add `if(trk)` to avoid `MarlinTrk::IMarlinTrack* trk = fit( *icv ) ;` return null pointer.
* correct the wrong-input detector parameters for CRD_o1_v04
  * because the parameters of 3rd and 4th layers are confused, the silicon trackers are defined not as expected. Let parameter use right sources. 
* Besides bug fix, also
  * fix hardcode for fit-direction: `marlinTrk->initialise( *pre_fit, bfield_z, IMarlinTrack::backward ) ;` => `marlinTrk->initialise( *pre_fit, bfield_z, fit_backwards ) ;` , here fit_backwards is as parameter, but IMarlinTrack::backward is a fixed.
  * add cut to produce SimTrackerHit/TrackerHit to avoid unexpected hit: before modifying, electron will produce many hits from radiation photon.
    * remove hits while passing neutral particles
    * add energy cut to keep TrackerHit, and add smearing condition pull region (default 1000) to drop thos far away from SimTrackerHit. Default is close not to apply the requirement, since 1000 is large enough. 
  * fix time lost while PlanarDigiAlg digitize SimTrackerHit to TrackerHit

### CODE REVIEW

File: Simulation/DetSimSD/src/TrackerCombineSensitiveDetector.cpp [hunk 1]
```diff
@@ -23,6 +23,8 @@ void TrackerCombineSensitiveDetector::Initialize(G4HCofThisEvent* HCE){
 
 G4bool TrackerCombineSensitiveDetector::ProcessHits(G4Step* step, G4TouchableHistory*){
   dd4hep::sim::Geant4StepHandler h(step);
+  if (fabs(h.trackDef()->GetPDGCharge()) < 0.01) return true;
```
Comment (mirguest) [inline Simulation/DetSimSD/src/TrackerCombineSensitiveDetector.cpp:26]: Hi @fucd,
Is this expected?

File: Utilities/KalTest/src/kaltracklib/TKalDetCradle.cxx [hunk 2]
```diff
@@ -45,6 +46,19 @@ fDone(kFALSE), fIsClosed(kFALSE)
...
+  TIter next(this);
+  TObject *mlp = 0;
+  while ((mlp = next())) {
+    TAttElement* det = const_cast<TAttElement*>(&(dynamic_cast<TAttElement *>(mlp)->GetParent(kFALSE)));
+    if(det_nelement.find(det)!=det_nelement.end()) det_nelement[det]++;
+    else det_nelement[det] = 1;
+  }
+  for (auto it : det_nelement) {
+    delete it.first;
+  }
```
Comment (mirguest) [inline Utilities/KalTest/src/kaltracklib/TKalDetCradle.cxx:61]: Can you explain a bit about this code?

**Review by mirguest — COMMENTED** [review]
It would be better to update the description, so people could understand the PR better. Thank you.

File: Simulation/DetSimSD/src/TrackerCombineSensitiveDetector.cpp (hunk 1 above)
Comment (fucd) [reply Simulation/DetSimSD/src/TrackerCombineSensitiveDetector.cpp:26]: @mirguest Yes, this is one step same as CDR. More reasonable condition will be test/considered continuely.

**Review by mirguest — APPROVED** [review]
Looks good to me. Thank you.

**Pull request merged**